torchvision/prototype/models/depth/stereo/raft_stereo.py

```diff
@@ -0,0 +1,746 @@
+        # see: https://github.com/princeton-vl/RAFT-Stereo/blob/main/core/raft_stereo.py#L35-L37
+        self.shared_base = shared_base
+        if shared_base:
+            self.residual_block = block(base_dim, base_dim, norm_layer=nn.InstanceNorm2d, stride=1)
+            self.conv = nn.Conv2d(base_dim, output_dim, kernel_size=3, padding=1)
+        else:
+            self.conv = nn.Conv2d(base_dim, output_dim, kernel_size=1)
+
+    def forward(self, x: Tensor) -> Tensor:
+        x = self.base_encoder(x)
+        if self.shared_base:
+            x = self.residual_block(x)
+        x = self.conv(x)
+        return x
+
+
+class MultiLevelContextEncoder(nn.Module):
+    def __init__(
+        self,
+        base_encoder: nn.Module,
+        output_dim: int = 256,
+        out_with_blocks: List[int] = [1, 1, 0],
+        block: Callable[..., nn.Module] = ResidualBlock,
+    ):
+        super().__init__()
+        self.num_level = len(out_with_blocks)
+        self.base_encoder = base_encoder
+        self.base_downsampling_ratio = base_encoder.downsampling_ratio
+        base_dim = base_encoder.output_dim
+
+        # Layer to output hidden_state and context separately (each produce output_dim//2 dims)
+        self.out_hidden_states = nn.ModuleList(
+            [
+                self._make_out_layer(base_dim, output_dim // 2, with_block=with_block, block=block)
+                for with_block in out_with_blocks
+            ]
+        )
+        self.out_contexts = nn.ModuleList(
+            [
+                self._make_out_layer(base_dim, output_dim // 2, with_block=with_block, block=block)
+                for with_block in out_with_blocks
+            ]
+        )
+
+        self.downsamplers = nn.ModuleList(
+            [self._make_downsampler(block, base_dim, base_dim) for i in range(1, self.num_level)]
+        )
+
+    def _make_out_layer(self, in_channels, out_channels, with_block=1, block=ResidualBlock):
+        conv_layer = nn.Conv2d(in_channels, out_channels, kernel_size=3, padding=1)
+        if with_block:
+            block_layer = block(in_channels, in_channels, norm_layer=nn.BatchNorm2d, stride=1)
+            return nn.Sequential(block_layer, conv_layer)
+        else:
+            return conv_layer
+
+    def _make_downsampler(self, block, in_channels, out_channels):
+        block1 = block(in_channels, out_channels, norm_layer=nn.BatchNorm2d, stride=2)
+        block2 = block(out_channels, out_channels, norm_layer=nn.BatchNorm2d, stride=1)
+        return nn.Sequential(block1, block2)
+
+    def forward(self, x: Tensor) -> List[Tensor]:
+        x = self.base_encoder(x)
+        outs = [torch.cat([self.out_hidden_states[0](x), self.out_contexts[0](x)], dim=1)]
+        for i in range(0, self.num_level - 1):
+            x = self.downsamplers[i](x)
+            outs.append(torch.cat([self.out_hidden_states[i + 1](x), self.out_contexts[i + 1](x)], dim=1))
```

Comment:
Nit: You can avoid having the `torch.cat` twice here by doing minor refactoring.